test/fixtures/es-module-loaders/mock-loader.mjs

```diff
@@ -0,0 +1,239 @@
+// module location is mocked. Since a single URL can be mocked multiple
+// times but it cannot be removed from the cache `mock-facade:` URLs have a
+// form of mock-facade:$VERSION:$REPLACING_URL with the parameters being URL
+// percent encoded every time a module is resolved
+//
+// NOTE: due to ESM spec, once a specifier has been resolved in a source text
+//       it cannot be changed. so things like the following DO NOT WORK
+//
+// ```mjs
+// import mock from 'node:mock';
+// mock('file:///app.js', {x:1});
+// const namespace1 = await import('file:///app.js');
+// namespace1.x; // 1
+// mock('file:///app.js', {x:2});
+// const namespace2 = await import('file:///app.js');
+// namespace2.x; // STILL 1, because this source text already set the specifier
+//               // for 'file:///app.js', a different specifier that resolves
+//               // to that could still get a new namespace though
+// assert(namespace1 === namespace2);
+// ```
+
+/**
+ * FIXME: this is a hack to workaround loaders being
+ * single threaded for now, just ensures that the MessagePort drains
+ */
+function doDrainPort() {
+  let msg;
+  while (msg = receiveMessageOnPort(preloadPort)) {
+    onPreloadPortMessage(msg.message);
+  }
+}
+
+/**
+ * @param param0 message from the application context
+ */
+function onPreloadPortMessage({
+  mockVersion, resolved, exports
+}) {
+  currentMockVersion = mockVersion;
+  mockedModuleExports.set(resolved, exports);
+}
+let preloadPort;
+export function globalPreload({port}) {
+  // Save the communication port to the application context to send messages
+  // to it later
+  preloadPort = port;
+  // Every time the application context sends a message over the port
+  port.on('message', onPreloadPortMessage);
+  // This prevents the port that the Loader/application talk over
+  // from keeping the process alive, without this, an application would be kept
+  // alive just because a loader is waiting for messages
+  port.unref();
+
+  const insideAppContext = (getBuiltin, port, setImportMetaCallback) => {
+    /**
+     * This is the Map that saves *all* the mocked URL -> replacement Module
+     * mappings
+     * @type {Map<string, {namespace, listeners}>}
+     */
+    let mockedModules = new Map();
+    let mockVersion = 0;
+    /**
+     * This is the value that is placed into the `node:mock` default export
+     * 
+     * @example
+     * ```mjs
+     * import mock from 'node:mock';
+     * const mutator = mock('file:///app.js', {x:1});
+     * const namespace = await import('file:///app.js');
+     * namespace.x; // 1;
+     * mutator.x = 2;
+     * namespace.x; // 2;
+     * ```
+     * 
+     * @param {string} resolved an absolute URL HREF string
+     * @param {object} replacementProperties an object to pick properties from
+     *                                       to act as a module namespace
+     * @returns {object} a mutator object that can update the module namespace
```

Comment:
```suggestion
     * @returns {Object} a mutator object that can update the module namespace
```

Reply:
I'm not sure what codebase you have been reading that uses lowercase for objects here, but when these are evaluated by our errors, the uppercase variant of this is generally preferred as it is faster because it gets evaluated first.

Reply:
lower case on purpose, since it doesn't inherit from Object.prototype. should not be changed.

Reply:
Agree with @bmeck here. It would be better still if we could define the shape of the object, but if it can be anything, `object` is better than `any`.